### PR TITLE
Allow octave to build with gfortran using MKL

### PR DESCRIPTION
**Pull request by mcuma**
By default the Intel compiler based MKL library is being used, this pull request modifies the Octave package to use the GNU based MKL library component, similarly to what was done for the R package at https://github.com/spack/spack/issues/8642

Speaking of this, it may be a good idea to do the same replacement for the PGI/Nvidia compiler, though, I presume that noone will be building Octave with PGI.

### CODE REVIEW

**Comment by mcuma** [comment]
This is my first pull request so apologies for the formatting problems. 

What's the process with fixing these? Can the approver do it or do I have to fix it first?

Also, are there any guidelines for the formatting?

Thanks,
MC

**Comment by alalazo** [comment]
You can see the issues by clinking on "Details" next to the failing test. In this case:
```
var/spack/repos/builtin/packages/octave/package.py:13: [E302] expected 2 blank lines, found 1
var/spack/repos/builtin/packages/octave/package.py:277: [E128] continuation line under-indented for visual indent
```

@mtmiller you may want to have a look and confirm changes are fine with you.

File: var/spack/repos/builtin/packages/octave/package.py [hunk 1]
```diff
@@ -270,6 +270,20 @@ def configure_args(self):
         else:
             config_args.append("--without-z")
 
+        if '^mkl' in spec and 'gfortran' in self.compiler.fc:
```
Comment (mtmiller) [inline var/spack/repos/builtin/packages/octave/package.py]: Wouldn't it be better to move this up to where these options are set first, around line 160?

Comment (mcuma) [reply var/spack/repos/builtin/packages/octave/package.py]: Yes, please. I followed the https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/r/package.py, but should have seen the BLAS config around line 160 in this package file.

Comment (mcuma) [reply var/spack/repos/builtin/packages/octave/package.py]: It took me some trying to get the formatting style right but it's looking OK now. I am not sure what's up with the failed codecov/project check, anyone knows how to fix that?

File: var/spack/repos/builtin/packages/octave/package.py [hunk 2]
```diff
@@ -164,6 +165,20 @@ def configure_args(self):
             "--with-lapack=%s" % spec['lapack'].libs.ld_flags
         ])
 
+        if '^mkl' in spec and 'gfortran' in self.compiler.fc:
```
Comment (mtmiller) [inline var/spack/repos/builtin/packages/octave/package.py:163]: This is still separate from the above block, wouldn't it be better to combine into one block that sets the blas and lapack options at once?

Comment (mcuma) [reply var/spack/repos/builtin/packages/octave/package.py:163]: Yes, thanks for noticing it. I am following https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/r/package.py and they have it that way too. I'll make the change in a moment.

Comment (mcuma) [reply var/spack/repos/builtin/packages/octave/package.py:163]: Just pushed the update with the combined block, sorry for the delay, got sidetracked.

Comment (mtmiller) [reply var/spack/repos/builtin/packages/octave/package.py:163]: Excellent :+1:

File: var/spack/repos/builtin/packages/py-ipython/package.py [hunk 3]
```diff
@@ -44,7 +44,7 @@ class PyIpython(PythonPackage):
     depends_on('py-prompt-toolkit@2.0.0:2.0.999', when='@7.5.0', type=('build', 'run'))
     depends_on('py-prompt-toolkit@2.0.0:2.999,3.0.2:3.0.999', when='@7.18:', type=('build', 'run'))
     depends_on('py-pygments', type=('build', 'run'))
-    depends_on('py-backcall', type=('build', 'run'), when='@7.5.0:')
+    depends_on('py-backcall', type=('build', 'run'), when='@7.3.0:')
```
Comment (mtmiller) [inline var/spack/repos/builtin/packages/py-ipython/package.py]: This looks unrelated, accidentally fell into this branch?

**Review by mtmiller — CHANGES_REQUESTED** [review]
Except for the ipython change, which looks like an accident to me, this looks perfect now, thank you for working so hard on this!

I'm sorry it took so long, I somehow missed the latest update comments and this fell competely off my radar.

File: var/spack/repos/builtin/packages/py-ipython/package.py (hunk 3 above)
Comment (mcuma) [reply var/spack/repos/builtin/packages/py-ipython/package.py]: Yes, the ipython is accidental, I must have made an update for that in a wrong branch. Sorry about that and thanks for accepting the Octave pull request. Do I need to undo the ipython change by updating this pull request or can you do it?

Comment (mtmiller) [reply var/spack/repos/builtin/packages/py-ipython/package.py]: I don't think I have access to push to your branch, it would be easier if you could make the change, either with a rebase or a new commit, I don't think it matters here.

**Comment by mcuma** [comment]
Thanks Mike, I hopefully removed the ipython change and the CI checks should be starting so I'll check in a bit to see if they passed and ping back.

**Comment by mcuma** [comment]
Hmm, looks like you now may need to approve the change before the CI check can start, so, please, check that when you get a chance.

**Review by mtmiller — DISMISSED** [review]
Looks great to me

**Comment by mtmiller** [comment]
@alalazo I've tested and reviewed this as maintainer, can you approve this as developer?

**Review by alalazo — APPROVED** [review]

**Comment by alalazo** [comment]
@mcuma can you please rebase and push?

**Comment by mcuma** [comment]
@alalazo, not sure I need to rebase and push as the changes in this pull request seem to be in the develop branch now. Sorry I don't deal with git in this capacity much.

**Comment by alalazo** [comment]
@mcuma I'll close this PR then, even though this is what I see in `develop`:
https://github.com/spack/spack/blob/5fa075f5b41e16ae04ee09824f8b7a43847fc106/var/spack/repos/builtin/packages/octave/package.py#L162-L166

**Comment by mtmiller** [comment]
@mcuma Unfortunately, I think you're looking at your own `develop` branch, your changes have not been merged yet. It would be really helpful if you'd consider rebasing your changes from scratch so they can be merged and everyone can benefit.

**Comment by mcuma** [comment]
OK, thanks, I created a new pull request, now knowing that the best approach is to have a separate branch for each pull request. I was looking at a different package when communicating a few days ago.
